mypyc/genops.py

```diff
@@ -1653,6 +1653,139 @@ def visit_try_stmt(self, t: TryStmt) -> Value:
 
         self.activate_block(exit_block)
 
+    def visit_try_body(self, t: TryStmt) -> None:
+        if t.handlers:
+            self.visit_try_except_stmt(t)
+        else:
+            self.accept(t.body)
+
+    class TryFinallyNonlocalControl(NonlocalControl):
+        def __init__(self, target: BasicBlock, ret_reg: Register) -> None:
+            self.target = target
+            self.ret_reg = ret_reg
+
+        def gen_break(self, builder: 'IRBuilder') -> None:
+            assert False, "unimplemented"
+
+        def gen_continue(self, builder: 'IRBuilder') -> None:
+            assert False, "unimplemented"
+
+        def gen_return(self, builder: 'IRBuilder', value: Value) -> None:
+            builder.add(Assign(self.ret_reg, value))
+            builder.add(Goto(self.target))
+
+    class FinallyNonlocalControl(CleanupNonlocalControl):
+        """Nonlocal control for finally blocks.
+
+        Just makes sure that sys.exc_info always gets restored when we
+        leave and the return register is decrefed if it isn't null.
+        """
+        def __init__(self, outer: NonlocalControl, ret_reg: Value, saved: Value) -> None:
+            super().__init__(outer)
+            self.ret_reg = ret_reg
+            self.saved = saved
+
+        def gen_cleanup(self, builder: 'IRBuilder') -> None:
+            # Do an error branch on the return value register, which
+            # may be undefined. This will allow it to be properliy
+            # decrefed if it is not null. This is kind of a hack.
+            target = BasicBlock()
+            builder.add(Branch(self.ret_reg, target, target, Branch.IS_ERROR))
+            builder.activate_block(target)
+
+            # Restore the old exc_info
+            # Don't bother plumbing a line through because it can't fail
+            target, cleanup = BasicBlock(), BasicBlock()
+            builder.add(Branch(self.saved, target, cleanup, Branch.IS_ERROR))
+            builder.activate_block(cleanup)
+            builder.primitive_op(restore_exc_info_op, [self.saved], -1)
+            builder.goto_and_activate(target)
+
+    def visit_try_finally_stmt(self, try_body: TryStmt, finally_body: Block) -> None:
```

Comment:
This method is pretty long. It would probably read better if it was split into multiple smaller methods.